design/proposals/modular-components.md

```diff
@@ -0,0 +1,163 @@
+   can be locally cached once per k8s API version in `lib`.
+3. A component should not be uniquely defined by it's base file name.
+   [More details soon]
+4. Handle environment specifications in `app.yaml` and retire `spec.json`.
+   Multiple files and areas for configuration can cause confusions for users.
+   In order to keep configurations simple, I propose that all ksonnet-specific
+   _app-level_ configurations reside in the `app.yaml`.
+
+`app.yaml` would add an _environments_ section as follows:
+
+```
+apiVersion: 0.0.1
+kind: ksonnet.io/app
+name: config
+version: 0.0.1
+registries:
+  incubator:
+    gitVersion:
+      commitSha: ea3408d44c2d8ea4d321364e5533d5c60e74bce0
+      refSpec: master
+    protocol: github
+    uri: github.com/ksonnet/parts/tree/master/incubator
+environments:
+  dev:
+    k8sVersion: 1.9.0
+    server: https://my-dev-server.com
+    namespaces:
+      - dev-foo
+      - dev-bar
+    targets:
+      - auth
+      - dev
+  staging:
+    k8sVersion: 1.8.0
+    server: https://my-staging-server.com
+    namespaces:
+      - staging
+    targets:
+      - redis.jsonnet
+      - auth
+      - prod
+```
+
+The _environments_ section operates similarly to the existing `spec.json` with
+the additions of:
+
+1. Support for multiple environment namespaces. Deployments will happen across
+   all the namespaces for that environment. If a user wishes to have separate
+   deployments to separate namespaces, these deployment groups should be
+   modeled as two separate environments.
+2. The introduction of _targets_. _Targets_ are the component targets that an
+   environment wishes to include in a deployment. _Target_ definitions are
+   equivalent to the directory or file name of the component.
+3. The introduction of _k8sVersion_, the Kubernetes version the environment
+   server is running on.
+
+Following the above `app.yaml` example, the users directory tree may look
+something like the following _(hidden files / directories are excluded)_:
+
+```
+my-ksonnet-app
+├── app.yaml
+├── components
+│   ├── auth
+│   │   ├── ca-secret.jsonnet
+│   │   └── tls-certificate.jsonnet
+│   ├── dev
+│   │   └── memcached.jsonnet
+│   ├── params.libsonnet
+│   ├── prod
+│   │   └── database
+│   │       ├── db-secret.jsonnet
+│   │       ├── db-service.yaml
+│   │       └── db-stateful.jsonnet
+│   └── redis.jsonnet
+├── environments
+│   ├── dev
+│   │   └── params.libsonnet
+│   └── prod
+│       └── params.libsonnet
+├── lib
+└── vendor
+```
+
+_Note_: There is no correlation between the environment _dev_ and the
+components directory _dev_; similarly, with _prod_. We are only calling it as
+such out of convenience. The user can call the components subdirectory whatever
+they desire, so long as it is specified as a _target_ in the `app.yaml` file
+for that environment.
+
+## User Experience
+
+### Backwards compatibility
+
+This change will be *not* be backwards-compatible due to the significant changes
```

Comment:
It could be possible to make this change backwards compatible if we version `app.yaml`

Reply:
The backwards compatibility is more of an issue of `<= 0.8.0 ks` expecting an environments directory with a `spec.json` file, `.metadata` folder etc.